src/Storages/Hive/HiveFile.cpp

```diff
@@ -79,15 +79,20 @@ Range createRangeFromParquetStatistics(std::shared_ptr<parquet::ByteArrayStatist
 
 std::optional<size_t> IHiveFile::getRows()
 {
-    if (!rows)
+    if (!has_init_rows)
+    {
+        std::lock_guard lock(mutex);
         rows = getRowsImpl();
```

Comment:
Should we check `has_init_rows` once again under the lock not to call `getRowsImpl` twice?

Reply:
You'r right